modules/grpc/common/schema/grpc-schema.hpp

```diff
@@ -75,26 +75,95 @@ struct NameValueTemplatePair
 struct Field
 {
   NameValueTemplatePair nv;
-  google::protobuf::FieldDescriptorProto::Type type;
   const google::protobuf::FieldDescriptor *field_desc;
 
-  Field(std::string name_, google::protobuf::FieldDescriptorProto::Type type_, LogTemplate *value_)
-    : nv(name_, value_), type(type_), field_desc(nullptr) {}
+  Field(LogTemplate *value_)
+    : nv("", value_) {}
 
   Field(const Field &a)
-    : nv(a.nv), type(a.type), field_desc(a.field_desc) {}
+    : nv(a.nv), field_desc(a.field_desc) {}
 
   Field &operator=(const Field &a)
   {
     nv = a.nv;
-    type = a.type;
     field_desc = a.field_desc;
 
     return *this;
   }
 
 };
 
+class ProtoSchemaProvider
+{
+public:
+  virtual bool init() = 0;
```

Comment:
The existence of a pure virtual `init()` method suggests that something may be wrong with the class hierarchy (the name sounds like a virtualized constructor).

Can we rename this method or fix the hierarchy so that constructors can do their work without dynamic dispatch?

Reply:
I believe this is the exact virtual init that we have for `LogPipe`s. The `ProtoSchemaProvider` implementations can have an initial empty state produced by the ctor, then grammar rules can fill them with values, then we have point where we want to initialize them. I intentionally named it `init()`, so it matched the `LogPipe`-like convention.

Reply:
I understand, thank you.